Mathlib/CategoryTheory/Limits/HasLimits.lean

```diff
@@ -0,0 +1,1355 @@
+    limit.lift G t ≫ (HasLimit.isoOfNatIso w).inv =
+      limit.lift F ((Cones.postcompose w.inv).obj _) :=
+  IsLimit.lift_comp_conePointsIsoOfNatIso_inv _ _ _
+#align category_theory.limits.has_limit.lift_iso_of_nat_iso_inv CategoryTheory.Limits.HasLimit.lift_isoOfNatIso_inv
+
+/-- The limits of `F : J ⥤ C` and `G : K ⥤ C` are isomorphic,
+if there is an equivalence `e : J ≌ K` making the triangle commute up to natural isomorphism.
+-/
+def HasLimit.isoOfEquivalence {F : J ⥤ C} [HasLimit F] {G : K ⥤ C} [HasLimit G] (e : J ≌ K)
+    (w : e.functor ⋙ G ≅ F) : limit F ≅ limit G :=
+  IsLimit.conePointsIsoOfEquivalence (limit.isLimit F) (limit.isLimit G) e w
+#align category_theory.limits.has_limit.iso_of_equivalence CategoryTheory.Limits.HasLimit.isoOfEquivalence
+
+@[simp]
+theorem HasLimit.isoOfEquivalence_hom_π {F : J ⥤ C} [HasLimit F] {G : K ⥤ C} [HasLimit G]
+    (e : J ≌ K) (w : e.functor ⋙ G ≅ F) (k : K) :
+    (HasLimit.isoOfEquivalence e w).hom ≫ limit.π G k =
+      limit.π F (e.inverse.obj k) ≫ w.inv.app (e.inverse.obj k) ≫ G.map (e.counit.app k) := by
+  simp only [HasLimit.isoOfEquivalence, IsLimit.conePointsIsoOfEquivalence_hom]
+  dsimp
+  simp
+#align category_theory.limits.has_limit.iso_of_equivalence_hom_π CategoryTheory.Limits.HasLimit.isoOfEquivalence_hom_π
+
+@[simp]
+theorem HasLimit.isoOfEquivalence_inv_π {F : J ⥤ C} [HasLimit F] {G : K ⥤ C} [HasLimit G]
+    (e : J ≌ K) (w : e.functor ⋙ G ≅ F) (j : J) :
+    (HasLimit.isoOfEquivalence e w).inv ≫ limit.π F j = limit.π G (e.functor.obj j) ≫ w.hom.app j :=
+  by
+  simp only [HasLimit.isoOfEquivalence, IsLimit.conePointsIsoOfEquivalence_hom]
+  dsimp
+  simp
+#align category_theory.limits.has_limit.iso_of_equivalence_inv_π CategoryTheory.Limits.HasLimit.isoOfEquivalence_inv_π
+
+section Pre
+
+variable (F) [HasLimit F] (E : K ⥤ J) [HasLimit (E ⋙ F)]
+
+/-- The canonical morphism from the limit of `F` to the limit of `E ⋙ F`.
+-/
+def limit.pre : limit F ⟶ limit (E ⋙ F) :=
+  limit.lift (E ⋙ F) ((Limit.cone F).whisker E)
+#align category_theory.limits.limit.pre CategoryTheory.Limits.limit.pre
+
+@[reassoc (attr := simp)]
+theorem limit.pre_π (k : K) : limit.pre F E ≫ limit.π (E ⋙ F) k = limit.π F (E.obj k) := by
+  erw [IsLimit.fac]
+  rfl
+#align category_theory.limits.limit.pre_π CategoryTheory.Limits.limit.pre_π
+
+@[simp]
+theorem limit.lift_pre (c : Cone F) :
+    limit.lift F c ≫ limit.pre F E = limit.lift (E ⋙ F) (c.whisker E) := by ext; simp
+#align category_theory.limits.limit.lift_pre CategoryTheory.Limits.limit.lift_pre
+
+variable {L : Type u₃} [Category.{v₃} L]
+
+variable (D : L ⥤ K) [HasLimit (D ⋙ E ⋙ F)]
+
+-- Porting note: added because instance not automatically synthesized
+instance [h : HasLimit (D ⋙ E ⋙ F)] : HasLimit ((D ⋙ E) ⋙ F) where 
+  exists_limit := by 
+    rw [Functor.assoc]
+    exact h.exists_limit
+instance [h : HasLimit ((D ⋙ E) ⋙ F)] : HasLimit (D ⋙ E ⋙ F) where 
+  exists_limit := by 
+    rw [← Functor.assoc]
+    exact h.exists_limit
```

Comment:
I think a Type class inference expert needs to look at this. @gebner is this going to cause problems? It looks like there's some loop going on but maybe it's okay.

Reply:
What goes wrong if you remove these two instances?

Reply:
Everything below it goes red because it cannot synthesize them. Since Lean cannot currently go from (f \circ g) \circ h to f \circ (g \circ h), inside `HasLimit` I don't think it can go back. But probably the safest thing is to add them to the declarations manually.

Reply:
BTW, since associativity of functor composition is true by definition, these instances could be replaced with:
```lean
instance [h : HasLimit (D ⋙ E ⋙ F)] : HasLimit ((D ⋙ E) ⋙ F) := h
instance [h : HasLimit ((D ⋙ E) ⋙ F)] : HasLimit (D ⋙ E ⋙ F) := h
```

But we should figure out whether these instances will cause issues as Chris suggested.

Reply:
> BTW, since associativity of functor composition is true by definition, these instances could be replaced with:
> 
> ```lean
> instance [h : HasLimit (D ⋙ E ⋙ F)] : HasLimit ((D ⋙ E) ⋙ F) := h
> instance [h : HasLimit ((D ⋙ E) ⋙ F)] : HasLimit (D ⋙ E ⋙ F) := h
> ```
> 
> But we should figure out whether these instances will cause issues as Chris suggested.

This works. But this is the best I can do to 
```lean
theorem limit.pre_pre [h : HasLimit (D ⋙  E ⋙  F)] : haveI : HasLimit ((D ⋙  E) ⋙  F) := h;
    limit.pre F E ≫ limit.pre (E ⋙ F) D = limit.pre F (D ⋙ E) := by
  haveI : HasLimit ((D ⋙  E) ⋙  F) := h
  ext j; erw [assoc, limit.pre_π, limit.pre_π, limit.pre_π]; rfl
```

Reply:
I've made these changes and removed the global instances.

Reply:
Regardless of looping, I've concluded this is a dumb thing to do.